### PR TITLE
Updates to schema for 3.0.0 to allow an array of processing elements

### PR DESCRIPTION
**Pull request by brian-r-calder**
The current schema requires that the processing elements are a dictionary of set stages of processing.  This makes it difficult to extend for different stages, and doesn't allow for timestamping to indicate the sequence of events.

The proposed modification makes this into an array, and adds timestamps and recognition strings to the elements so that ordering is possible, and parsers can determine what to expect in the dictionary based on the recognition text.

In addition:

1. "Waterline" has been added as a potential solution for vertical reference.
2. Mandatory requirements are set for processing elements to ensure that the minimal information required is enforced when the metadata is validated.

### CODE REVIEW

File: csbschema/__init__.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 from csbschema import validators
 
 
-__version__ = '1.0.1'
+__version__ = '1.0.2.dev0'
```
Comment (selimnairb) [inline csbschema/__init__.py:7]: this should be changed to `1.0.2` before merging.

**Comment by selimnairb** [comment]
@brian-r-calder I believe this is ready to merge (after updating the version number of course). Let me know if you see anything else that needs to be addressed before moving forward.

**Review by selimnairb — APPROVED** [review]

**Comment by selimnairb** [comment]
@brian-r-calder Ready to merge.

**Review by georgiannazelenak — CHANGES_REQUESTED** [review]
DCDB systems do not use the "processing"/"lineage" element and therefore will not be affected by making it an array.  As previously mentioned, the DCDB cannot currently support v3bis due to the location of the uniqueID field.  Our current pipeline would also miss the ship name and instrument fields, both of which we present on map viewers when provided.

During general review of the schema in its entirety I noted that many of the field names, and in some cases the order presented, vary from the example file currently provided (https://www.ngdc.noaa.gov/iho/SampleCSBFileFormats.pdf) and referenced by trusted nodes.  This includes required fields, leading me to suspect that most current submissions would fail validation against this schema.  Two of the differences in CSB-schema-3_0_0-2023-02.json would negatively impact DCDB ingest systems:

1. DCDB systems use the UniqueID from the metadata.  Systems look specifically for the "uniqueID" field, therefore "uniqueVesselID" would need to be renamed.  DCDB systems are also set up to look for this field within the "platform" element (with a secondary search at the root level if needed).  Having this field within "trustedNode" would cause file rejection.  Note that the expression for the uniqueID in this repo is also more restrictive than that used by the DCDB ([0-9a-zA-Z]{2,}(-[0-9a-zA-Z]{2,})+).  The specification in this repo is the gold standard, but we have found that it is not consistently used.
2. When provided, we use the "sensors" element to populate "instrument" on the DCDB viewer.   DCDB systems would pass over "sensorDescriptions", can this be renamed  "sensors"?

Many other differences (orgName vs. providerOrganizationName, "trustedNode" vs. "providerContactPoint", formatting of the CRS element, etc) do not impact the DCDB directly.  However the differences mean that this schema is not representative of files as we receive them today.  I would be happy to provide line-by-line feedback, is that best done here or better handled as a new issue as it is beyond the scope of the changes where review was requested?  Note I don't object to the schema as presented here; much of what's presented is more logical than our current examples.  However, it seems best handled as a new version (a larger discussion).

**Comment by selimnairb** [comment]
After meeting with @georgiannazelenak and @brian-r-calder the following course of action for this pull request was agreed upon:

- The current 3.0.0 schema will be renamed 3.1.0
- Version 3.1.0 will be modified as follows to ensure backward compatibility with DCDB ingest processes:
  - Duplicate `/properties/trustedNode/uniqueVesselID` as `/properties/platform/uniqueID`
  - Rename `/properties/platform/sensorDescriptions` to `/properties/platform/sensors`

Then, in a separate pull request to follow shortly after this pull request, a "new" 3.0.0 schema will be created that validates the following examples:
- https://www.ngdc.noaa.gov/iho/SampleCSBFileFormats.pdf 
- https://www.ngdc.noaa.gov/iho/example_csb_geojson_file.geojson

@georgiannazelenak will review the 3.1.0 changes to the current pull request, and will be a reviewer on the subsequent "new" 3.0.0 pull request.

Please let me know if I have omitted or mis-represented any pertinent details. Thanks!

**Comment by selimnairb** [comment]
> After meeting with @georgiannazelenak and @brian-r-calder the following course of action for this pull request was agreed upon:
> 
> * The current 3.0.0 schema will be renamed 3.1.0
> * Version 3.1.0 will be modified as follows to ensure backward compatibility with DCDB ingest processes:
>   
>   * Duplicate `/properties/trustedNode/uniqueVesselID` as `/properties/platform/uniqueID`
>   * Rename `/properties/platform/sensorDescriptions` to `/properties/platform/sensors`
> 
> Then, in a separate pull request to follow shortly after this pull request, a "new" 3.0.0 schema will be created that validates the following examples:
> 
> * https://www.ngdc.noaa.gov/iho/SampleCSBFileFormats.pdf
> * https://www.ngdc.noaa.gov/iho/example_csb_geojson_file.geojson
> 
> @georgiannazelenak will review the 3.1.0 changes to the current pull request, and will be a reviewer on the subsequent "new" 3.0.0 pull request.
> 
> Please let me know if I have omitted or mis-represented any pertinent details. Thanks!

The above comment is addressed in ffc54fa13809780ba82520d7e6b758a06e043a0b.

**Comment by selimnairb** [comment]
@brian-r-calder @georgiannazelenak Let me know if this is ready to merge. (this should be merged before branch 4-create-back-port-300-schema).

File: docs/IHO/b12_v3_1_0_example-noprocessing.json [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      "convention": "GeoJSON CSB 3.1",
+      "dataLicense": "CC0 1.0",
+      "providerLogger": "Rose Point ECS",
+      "providerLoggerVersion": "1.0",
+      "navigationCRS": "EPSG:4326",
+      "verticalReferenceOfDepth": "Transducer",
+      "vesselPositionReferencePoint": "GNSS"
+    },
+    "platform": {
+      "type": "Private vessel",
```
Comment (georgiannazelenak) [inline docs/IHO/b12_v3_1_0_example-noprocessing.json:23]: Can "uniqueID": "SEAID-e8c469f8-df38-11e5-b86d-9a79f06e9478" be added here for consistency with other examples and backwards compatibility with DCDB?

**Review by georgiannazelenak — CHANGES_REQUESTED** [review]
Thanks for your patience.  The only suggestion from DCDB is to add "uniqueID" in the "platforms" element in docs/IHO/b12_b3_1_0_example-noprocessing.json for consistency with other examples and backwards compatibility with DCDB ingest systems.

File: docs/IHO/b12_v3_1_0_example-noprocessing.json (hunk 2 above)
Comment (selimnairb) [reply docs/IHO/b12_v3_1_0_example-noprocessing.json:24]: @georgiannazelenak Good catch. This has been addressed in 507b0979fcd5cda1d8c33b6e8cd9688eca0f0150.

**Review by georgiannazelenak — APPROVED** [review]

**Comment by selimnairb** [comment]
@brian-r-calder Given the completed review by @georgiannazelenak I believe this is ready for merging.

**Pull request merged**